src/wp-includes/class-wp-query.php

```diff
@@ -3528,8 +3529,17 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
+				update_post_cache( $this->posts );
+			}
```

Comment:
In the example provided on the ticket the issue was that `$this->posts` may not contain a full post object.

I think it's best to cache the entire object reqardless using something like:

```php
if ( $q['cache_results'] ) {
        if ( $id_query_is_cacheable ) {
                update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
        } else {
                $post_ids = wp_list_pluck( $this->posts, 'ID' );
                _prime_post_caches( $post_ids, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
        }
}
```

`cache_results` has always been about the main query rather than priming the posts -- the current change will end up priming the posts as each `get_post()` is called so will take it from 1 query to potentially 10.

Reply:
I think this is a semantics discussion: We have two options here:
* Either we interpret the `cache_results` flag as "cache the query results", in which case the current way this PR is implemented makes sense. We will only cache the individual posts if their data is part of the query results, effectively getting this post caching "for free".
* Or we interpret the `cache_results` flag to also encompass "cache the query results and post results", in which case something like what you outlined @peterwilsoncc makes sense. The drawback would be that WP would run another DB query, so at that point I would question why it's even possible to query fewer post fields, if in the end all post fields are queried anyway.

I'm personally leaning towards the first option.

Reply:
@spacedmonkey @felixarntz Priming the entire object cache fixes the regression introduced in 5a497225b286964ff9302a21272731172840ed05 for WP 6.2.

Prior to that commit, the posts were primed using the IDs and an additional database query, the commit changed it so the posts were primed using the selected post data. What we (it looks the like three of us worked on it) missed in that change was when the data was filtered in the way reported in this ticket.

For a long time the individual post caches were primed regardless of the `cache_results` setting so I think it's best to continue doing that.

> ...at that point I would question why it's even possible to query fewer post fields, if in the end all post fields are queried anyway.

I don't disagree but as it's being the case for so long I'm hesitant to change it. I worry that instead of six database queries we'd end up with posts making 15 database queries because they don't realise the posts are filtered.

Reply:
@peterwilsoncc How about a new parameter that allow developers to opt-in/out of store post objects. That way we do not have to guess if this is the desired behaviour or not. 

I agree with @felixarntz, if are trying avoid loading, say large "post_content" field, then loading them anyway using `_prime_post_caches` will have negative effects. 

It is unclear to all of us, then it will be unclear to developers as well. 

Reminder, we don't need to prime post caches for the loop. The `$post` global comes from array stored in property. 

https://github.com/WordPress/wordpress-develop/blob/60d2d8ac75c4350a45e4c6970c7d0295b757f1f0/src/wp-includes/class-wp-query.php#L3669-L3670

https://github.com/WordPress/wordpress-develop/blob/60d2d8ac75c4350a45e4c6970c7d0295b757f1f0/src/wp-includes/class-wp-query.php#L3621-L3628

Not priming these most likely would not result in 15 database queries.

Reply:
@spacedmonkey After reviewing the logic again more closely, I found two related problems which may cause `$this->posts` to not contain full post objects, see https://github.com/WordPress/wordpress-develop/pull/5134#pullrequestreview-1633962867.

I think we will need to cater for that.

Separately, I agree now with @peterwilsoncc that for other cases we should continue to have `_prime_post_caches()` in an `else` for the case the post objects are incomplete, and here's why:
* The main reason to avoid the `_prime_post_caches()` call originally was that when `$this->posts` has the complete post objects, we can avoid a database query. This is still the case here.
* Calling `_prime_post_caches()` in an `else` will only be reached in edge-cases (e.g. where post fields or the actual results are filtered), but for those I agree that the post caches should continue to be primed, since otherwise it can lead to accidental excess database queries.

Adding back the `_prime_post_caches()` call in that scenario will not realistically have a notable negative performance impact, while ensuring that such edge-cases where query fields or query results are filtered still behave the same way as they did before. The original performance goal for why the main `_prime_post_caches()` call was removed is still maintained.